File: swarmx/core/engine.py
"""
SwarmX Engine — Core orchestration engine.

The SwarmEngine is the top-level coordinator that wires together the
event bus, task scheduler, provider registry, and agents. It is
responsible for:
  - Loading swarm definitions from config
  - Instantiating and managing agent lifecycles
  - Starting/stopping all subsystems
  - Providing a clean API for the CLI and programmatic use

Adapted from OpenClaw's Gateway pattern: the Gateway acts as the single
control plane for sessions, channels, tools, and events. The SwarmEngine
serves the same role for multi-agent orchestration.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any

from swarmx.core.agent import Agent, AgentConfig, AgentState
from swarmx.core.event_bus import Event, EventBus
from swarmx.core.provider import ProviderBase, ProviderConfig, ProviderRegistry
from swarmx.core.scheduler import Task, TaskScheduler

logger = logging.getLogger("swarmx.engine")


class SwarmEngine:
    """
    The SwarmX orchestration engine.

    Usage:
        engine = SwarmEngine()
        engine.register_provider("openai", OpenAIProvider(config))
        engine.add_agent(AgentConfig(name="researcher", provider="openai", ...))
        await engine.start()
        await engine.submit_task("Analyze the market")
        await engine.stop()
    """

    def __init__(self, event_bus: EventBus | None = None) -> None:
        self.event_bus = event_bus or EventBus()
        self.provider_registry = ProviderRegistry()
        self.scheduler = TaskScheduler(self.event_bus)

        self._agents: dict[str, Agent] = {}
        self._running = False

        # Register built-in provider classes
        self._register_builtin_providers()

        # Subscribe to engine-level events
        self.event_bus.subscribe(
            "agent.response.*",
            self._on_agent_response,
            subscriber_id="engine",
        )
        self.event_bus.subscribe(
            "agent.error",
            self._on_agent_error,
            subscriber_id="engine",
        )

    # ── Provider Management ─────────────────────────────────────────

    def _register_builtin_providers(self) -> None:
        """Register the built-in provider classes for lazy instantiation."""
        try:
            from swarmx.providers.openai_provider import OpenAIProvider
            self.provider_registry.register_class("openai", OpenAIProvider)
        except ImportError:
            pass

        try:
            from swarmx.providers.anthropic_provider import AnthropicProvider
            self.provider_registry.register_class("anthropic", AnthropicProvider)
        except ImportError:
            pass

        try:
            from swarmx.providers.google_provider import GoogleProvider
            self.provider_registry.register_class("google", GoogleProvider)
        except ImportError:
            pass

        try:
            from swarmx.providers.xai_provider import XAIProvider
            self.provider_registry.register_class("xai", XAIProvider)
        except ImportError:
            pass

    def register_provider(
        self, name: str, provider: ProviderBase | None = None, config: ProviderConfig | None = None
    ) -> None:
        """
        Register a provider by instance or by config (lazy creation).

        Args:
            name: Provider name that agents will reference.
            provider: A pre-configured provider instance.
            config: Config for lazy provider creation from registered classes.
        """
        if provider:
            self.provider_registry.register_instance(name, provider)
        elif config:
            self.provider_registry.create(name, config)
        else:
            raise ValueError("Must provide either a provider instance or config")

    # ── Agent Management ────────────────────────────────────────────

    def add_agent(
        self,
        config: AgentConfig,
        agent_class: type[Agent] | None = None,
    ) -> Agent:
        """
        Create and register an agent.

        Args:
            config: Agent configuration.
            agent_class: Optional custom agent class (defaults to base Agent).

        Returns:
            The created Agent instance.
        """
        cls = agent_class or Agent
        agent = cls(
            config=config,
            event_bus=self.event_bus,
            provider_registry=self.provider_registry,
        )
        self._agents[agent.agent_id] = agent
        logger.info("Agent added: %s", agent.agent_id)
        return agent

    def get_agent(self, agent_id: str) -> Agent | None:
        """Look up an agent by ID."""
        return self._agents.get(agent_id)

    def get_agents_by_name(self, name: str) -> list[Agent]:
        """Find all agents with a given name (there may be multiple instances)."""
        return [a for a in self._agents.values() if a.config.name == name]

    # ── Task Submission ─────────────────────────────────────────────

    async def submit_task(
        self,
        content: str,
        name: str = "",
        target_topic: str = "task.created",
        **kwargs: Any,
    ) -> str:
        """
        Submit a task to the swarm.

        This is the primary way to send work into the system.
        The task is published as an event and handled by subscribed agents.

        Returns:
            The task ID for tracking.
        """
        task = Task(
            name=name or content[:50],
            description=content,
            target_topic=target_topic,
            payload={"content": content, **kwargs},
        )
        return await self.scheduler.submit(task)

    async def broadcast(self, topic: str, payload: dict[str, Any]) -> None:
        """Publish an event to the bus directly."""
        event = Event(topic=topic, payload=payload, source="engine")
        await self.event_bus.publish(event)

    # ── Lifecycle ───────────────────────────────────────────────────

    async def start(self) -> None:
        """
        Start the SwarmX engine.

        Initializes all agents, starts the event bus, and begins scheduling.
        """
        if self._running:
            logger.warning("Engine already running")
            return

        logger.info("Starting SwarmX engine...")

        # Start event bus first
        await self.event_bus.start()

        # Initialize all agents
        init_tasks = [agent.initialize() for agent in self._agents.values()]
        await asyncio.gather(*init_tasks, return_exceptions=True)

        # Start scheduler
        await self.scheduler.start()

        self._running = True
        logger.info(
            "SwarmX engine started (%d agents, %d providers)",
            len(self._agents),
            len(self.provider_registry.available),
        )

    async def stop(self) -> None:
        """Gracefully shut down all subsystems."""
        if not self._running:
            return

        logger.info("Stopping SwarmX engine...")

        # Shutdown agents
        shutdown_tasks = [agent.shutdown() for agent in self._agents.values()]
        await asyncio.gather(*shutdown_tasks, return_exceptions=True)

        # Stop scheduler
        await self.scheduler.stop()

        # Stop event bus last
        await self.event_bus.stop()

        self._running = False
        logger.info("SwarmX engine stopped")

    # ── Internal Event Handlers ─────────────────────────────────────

    async def _on_agent_response(self, event: Event) -> None:
        """Handle agent response events (logging/routing)."""
        agent_id = event.payload.get("agent_id", "unknown")
        content = event.payload.get("content", "")
        logger.debug(
            "Agent response from %s: %s",
            agent_id,
            content[:100] + "..." if len(content) > 100 else content,
        )

    async def _on_agent_error(self, event: Event) -> None:
        """Handle agent error events."""
        agent_id = event.payload.get("agent_id", "unknown")
        logger.error("Agent error from %s: %s", agent_id, event.payload)

    # ── Introspection ───────────────────────────────────────────────

    @property
    def agents(self) -> dict[str, Agent]:
        """All registered agents."""
        return dict(self._agents)

    @property
    def is_running(self) -> bool:
        """Whether the engine is currently running."""
        return self._running

    def status(self) -> dict[str, Any]:
        """Return a summary of the engine's current state."""
        return {
            "running": self._running,
            "agents": {
                aid: {
                    "name": a.config.name,
                    "state": a.state.value,
                    "provider": a.config.provider,
                }
                for aid, a in self._agents.items()
            },
            "providers": self.provider_registry.available,
            "event_bus": self.event_bus.stats,
            "scheduler": {
                "pending": self.scheduler.pending_count,
                "running": self.scheduler.running_count,
            },
        }

    def __repr__(self) -> str:
        return (
            f"<SwarmEngine agents={len(self._agents)} "
            f"running={self._running}>"
        )

File: src/core/engine.ts
/**
 * Groklets Engine — Production-grade orchestration engine.
 *
 * Integrates all core subsystems:
 *   - EventBus (async event routing)
 *   - ProviderRegistry (model management + failover)
 *   - TaskScheduler (dependency-aware task queue)
 *   - SessionStore (persistent conversations)
 *   - UsageTracker (cost tracking)
 *   - ContextManager (token management)
 */

import { EventBus, createEvent, type SwarmEvent } from "./event-bus.js";
import { Agent, type AgentConfig } from "./agent.js";
import { ProviderRegistry, type ProviderBase, type ProviderConfig } from "./provider.js";
import { TaskScheduler, createTask } from "./scheduler.js";
import { SessionStore, type SessionStoreConfig } from "./session.js";
import { UsageTracker } from "./usage.js";
import { FailoverProvider, type FailoverConfig } from "./failover.js";
import { createLogger } from "../utils/logger.js";

const log = createLogger("Engine");

export interface EngineConfig {
    sessions?: SessionStoreConfig;
    failover?: FailoverConfig;
}

export class SwarmEngine {
    readonly eventBus: EventBus;
    readonly providerRegistry: ProviderRegistry;
    readonly scheduler: TaskScheduler;
    readonly sessionStore: SessionStore;
    readonly usageTracker: UsageTracker;

    private agents = new Map<string, Agent>();
    private _running = false;
    private shutdownHandlers: (() => void)[] = [];
    private engineConfig: EngineConfig;

    constructor(config?: EngineConfig, eventBus?: EventBus) {
        this.engineConfig = config ?? {};
        this.eventBus = eventBus ?? new EventBus();
        this.providerRegistry = new ProviderRegistry();
        this.scheduler = new TaskScheduler(this.eventBus);
        this.sessionStore = new SessionStore(config?.sessions);
        this.usageTracker = new UsageTracker();

        this.eventBus.subscribe("agent.response.*", (e) => this.onAgentResponse(e), "engine");
        this.eventBus.subscribe("agent.error", (e) => this.onAgentError(e), "engine");
    }

    // ── Providers ───────────────────────────────────────────────

    registerProvider(name: string, provider: ProviderBase): void;
    registerProvider(name: string, cls: new (c: ProviderConfig) => ProviderBase, config: ProviderConfig): void;
    registerProvider(
        name: string,
        providerOrClass: ProviderBase | (new (c: ProviderConfig) => ProviderBase),
        config?: ProviderConfig,
    ): void {
        if (config && typeof providerOrClass === "function") {
            const instance = new providerOrClass(config);
            this.providerRegistry.registerInstance(name, instance);
        } else {
            this.providerRegistry.registerInstance(name, providerOrClass as ProviderBase);
        }
        log.info(`Provider registered: ${name}`);
    }

    /**
     * Create a failover provider from multiple providers.
     * First provider is primary, rest are fallbacks.
     */
    registerFailover(name: string, providerNames: string[], config?: FailoverConfig): void {
        const providers = providerNames.map((n) => this.providerRegistry.get(n));
        const failover = new FailoverProvider(name, providers, config);
        this.providerRegistry.registerInstance(name, failover);
        log.info(`Failover registered: ${name} (${providerNames.join(" → ")})`);
    }

    // ── Agents ──────────────────────────────────────────────────

    addAgent(
        config: AgentConfig,
        AgentClass?: new (...args: ConstructorParameters<typeof Agent>) => Agent,
    ): Agent {
        const Cls = AgentClass ?? Agent;
        const agent = new Cls(config, this.eventBus, this.providerRegistry, this.sessionStore);
        this.agents.set(agent.agentId, agent);
        log.info(`Agent added: ${agent.agentId} → ${config.provider}`);
        return agent;
    }

    getAgent(id: string): Agent | undefined { return this.agents.get(id); }
    getAgentsByName(name: string): Agent[] {
        return [...this.agents.values()].filter((a) => a.config.name === name);
    }

    // ── Tasks ───────────────────────────────────────────────────

    async submitTask(
        content: string,
        opts?: { name?: string; targetTopic?: string; payload?: Record<string, unknown> },
    ): Promise<string> {
        const task = createTask({
            name: opts?.name ?? content.slice(0, 50),
            description: content,
            targetTopic: opts?.targetTopic ?? "task.created",
            payload: { content, ...opts?.payload },
        });
        return this.scheduler.submit(task);
    }

    async broadcast(topic: string, payload: Record<string, unknown>): Promise<void> {
        await this.eventBus.publish(createEvent({ topic, payload, source: "engine" }));
    }

    /**
     * Send a message from one agent to another via event bus.
     */
    async agentToAgent(fromAgent: string, toAgent: string, message: string): Promise<void> {
        await this.eventBus.publish(createEvent({
            topic: `agent.message.${toAgent}`,
            payload: { content: message, from: fromAgent },
            source: fromAgent,
        }));
    }

    // ── Lifecycle ───────────────────────────────────────────────

    async start(): Promise<void> {
        if (this._running) return;

        log.info("Starting Groklets engine...");

        await this.eventBus.start();

        const results = await Promise.allSettled(
            [...this.agents.values()].map((a) => a.initialize()),
        );
        const failed = results.filter((r) => r.status === "rejected");
        if (failed.length > 0) {
            log.warn(`${failed.length} agent(s) failed to initialize`);
        }

        await this.scheduler.start();
        this._running = true;
        this.trapSignals();

        log.info(`Engine running — ${this.agents.size} agents, ${this.providerRegistry.available.length} providers`);
    }

    async stop(): Promise<void> {
        if (!this._running) return;

        log.info("Shutting down...");

        this.shutdownHandlers.forEach((fn) => fn());
        this.shutdownHandlers = [];

        // Shutdown agents (saves sessions)
        await Promise.allSettled([...this.agents.values()].map((a) => a.shutdown()));

        // Save all sessions
        this.sessionStore.saveAll();

        await this.scheduler.stop();
        await this.eventBus.stop();

        this._running = false;

        // Log final usage
        const usage = this.getUsageSummary();
        log.info(`Engine stopped — ${usage.totalCalls} calls, ${usage.totalTokens} tokens, ${usage.totalCost}`);
    }

    private trapSignals(): void {
        const handler = async () => {
            log.info("Received shutdown signal");
            await this.stop();
            process.exit(0);
        };

        const sigint = () => { handler(); };
        const sigterm = () => { handler(); };

        process.on("SIGINT", sigint);
        process.on("SIGTERM", sigterm);

        this.shutdownHandlers.push(
            () => process.removeListener("SIGINT", sigint),
            () => process.removeListener("SIGTERM", sigterm),
        );
    }

    // ── Internal Handlers ───────────────────────────────────────

    private async onAgentResponse(event: SwarmEvent): Promise<void> {
        // Track usage from agent responses
        const usage = event.payload.usage as { promptTokens?: number; completionTokens?: number } | undefined;
        const model = event.payload.model as string | undefined;
        if (usage && model) {
            this.usageTracker.track(model, usage.promptTokens ?? 0, usage.completionTokens ?? 0);
        }
    }

    private async onAgentError(event: SwarmEvent): Promise<void> {
        log.error(`Agent error: ${event.payload.agentId} — ${event.payload.error}`);
    }

    // ── Introspection ───────────────────────────────────────────

    get allAgents() { return new Map(this.agents); }
    get isRunning() { return this._running; }

    getUsageSummary(): { totalCalls: number; totalTokens: number; totalCost: string; breakdown: Record<string, unknown> } {
        // Aggregate from all agents + engine-level
        let totalCalls = this.usageTracker.summary.calls;
        let totalTokens = this.usageTracker.summary.totalTokens;
        let totalCostUsd = this.usageTracker.summary.costUsd;

        for (const agent of this.agents.values()) {
            const agentUsage = agent.usageTracker.summary;
            totalCalls += agentUsage.calls;
            totalTokens += agentUsage.totalTokens;
            totalCostUsd += agentUsage.costUsd;
        }

        return {
            totalCalls,
            totalTokens,
            totalCost: `$${totalCostUsd.toFixed(4)}`,
            breakdown: this.usageTracker.breakdown(),
        };
    }

    status(): Record<string, unknown> {
        const agents: Record<string, unknown> = {};
        for (const [id, agent] of this.agents) {
            agents[id] = {
                name: agent.config.name,
                state: agent.state,
                provider: agent.config.provider,
                contextUsage: agent.contextUsage,
                usage: agent.usageTracker.summary,
                tools: agent.toolExecutor.registeredTools,
            };
        }
        return {
            running: this._running,
            agents,
            providers: this.providerRegistry.available,
            eventBus: this.eventBus.stats,
            scheduler: { pending: this.scheduler.pendingCount, running: this.scheduler.runningCount },
            sessions: this.sessionStore.stats(),
            usage: this.getUsageSummary(),
        };
    }

    /**
     * Full diagnostics — like `openclaw doctor`.
     */
    async doctor(): Promise<Array<{ check: string; status: "pass" | "warn" | "fail"; detail: string }>> {
        const checks: Array<{ check: string; status: "pass" | "warn" | "fail"; detail: string }> = [];

        // Engine
        checks.push({
            check: "Engine running",
            status: this._running ? "pass" : "fail",
            detail: this._running ? "Engine is running" : "Engine is not started",
        });

        // Providers
        for (const name of this.providerRegistry.available) {
            try {
                const provider = this.providerRegistry.get(name);
                if (provider.healthCheck) {
                    const ok = await provider.healthCheck();
                    checks.push({
                        check: `Provider: ${name}`,
                        status: ok ? "pass" : "fail",
                        detail: ok ? "Connected" : "Unreachable",
                    });
                } else {
                    checks.push({ check: `Provider: ${name}`, status: "warn", detail: "No health check available" });
                }
            } catch (err) {
                checks.push({ check: `Provider: ${name}`, status: "fail", detail: (err as Error).message });
            }
        }

        // Agents
        for (const [, agent] of this.agents) {
            const ctx = agent.contextUsage;
            checks.push({
                check: `Agent: ${agent.config.name}`,
                status: agent.state === "idle" ? "pass" : agent.state === "error" ? "fail" : "warn",
                detail: `State: ${agent.state}, Context: ${ctx.percent}% (${ctx.estimated}/${ctx.max} tokens)`,
            });
        }

        // Sessions
        const sessionStats = this.sessionStore.stats();
        checks.push({
            check: "Session store",
            status: "pass",
            detail: `${sessionStats.totalSessions} sessions, ${sessionStats.totalMessages} messages`,
        });

        return checks;
    }
}
